src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -82,41 +86,62 @@ object RuleUtils {
+          } else {
+            res
+          }
         }
+
+        val appendedBytesRatio = 1 - commonBytes / inputSourceFilesSizeInBytes.toFloat
+        val deletedBytesRatio = 1 - commonBytes / entry.sourceFilesSizeInBytes.toFloat
+
+        val deletedCnt = entry.sourceFileInfoSet.size - commonCnt
+        val isAppendAndDeleteCandidate = hybridScanDeleteEnabled && entry.hasLineageColumn &&
+          commonCnt > 0 &&
+          appendedBytesRatio < HyperspaceConf.hybridScanAppendedRatioThreshold(spark) &&
+          deletedBytesRatio < HyperspaceConf.hybridScanDeletedRatioThreshold(spark)
+
+        // For append-only Hybrid Scan, deleted files are not allowed.
+        lazy val isAppendOnlyCandidate = deletedCnt == 0 && commonCnt > 0 &&
+          appendedBytesRatio < HyperspaceConf.hybridScanAppendedRatioThreshold(spark)
+
+        val isCandidate = isAppendAndDeleteCandidate || isAppendOnlyCandidate
+        if (isCandidate) {
+          entry.setTagValue(plan, IndexLogEntryTags.COMMON_SOURCE_SIZE_IN_BYTES, commonBytes)
+
+          // If there is no change in source dataset, the index will be applied by
+          // transformPlanToUseIndexOnlyScan.
+          entry.setTagValue(
+            plan,
+            IndexLogEntryTags.HYBRIDSCAN_REQUIRED,
+            !(commonCnt == entry.sourceFileInfoSet.size && commonCnt == inputSourceFiles.size))
+        }
+        isCandidate
       }
+    }
 
-      val appendedBytesRatio = 1 - commonBytes / inputSourceFilesSizeInBytes.toFloat
-      val deletedBytesRatio = 1 - commonBytes / entry.sourceFilesSizeInBytes.toFloat
-
-      val deletedCnt = entry.sourceFileInfoSet.size - commonCnt
-      val isAppendAndDeleteCandidate = hybridScanDeleteEnabled && entry.hasLineageColumn &&
-        commonCnt > 0 &&
-        appendedBytesRatio < HyperspaceConf.hybridScanAppendedRatioThreshold(spark) &&
-        deletedBytesRatio < HyperspaceConf.hybridScanDeletedRatioThreshold(spark)
-
-      // For append-only Hybrid Scan, deleted files are not allowed.
-      lazy val isAppendOnlyCandidate = deletedCnt == 0 && commonCnt > 0 &&
-        appendedBytesRatio < HyperspaceConf.hybridScanAppendedRatioThreshold(spark)
-
-      val isCandidate = isAppendAndDeleteCandidate || isAppendOnlyCandidate
-      if (isCandidate) {
-        entry.setTagValue(plan, IndexLogEntryTags.COMMON_SOURCE_SIZE_IN_BYTES, commonBytes)
-
-        // If there is no change in source dataset, the index will be applied by
-        // transformPlanToUseIndexOnlyScan.
-        entry.setTagValue(
-          plan,
-          IndexLogEntryTags.HYBRIDSCAN_REQUIRED,
-          !(commonCnt == entry.sourceFileInfoSet.size && commonCnt == inputSourceFiles.size))
+    def prepareHybridScanCandidateSelection(
+        spark: SparkSession,
+        plan: LogicalPlan,
+        indexes: Seq[IndexLogEntry]): Unit = {
+      val curConfigs = Seq(
+        HyperspaceConf.hybridScanEnabled(spark).toString,
```

Comment:
Should we remove this? We can add `assert(HyperspaceConf.hybridScanEnabled(spark))` if needed.